src/Type/Php/ImplodeFunctionReturnTypeExtension.php

```diff
@@ -115,13 +115,17 @@ private function inferConstantType(ConstantArrayType $arrayType, ConstantStringT
 
 			$arrayValues = [];
 			foreach ($valueTypes as $valueType) {
-				if (!$valueType instanceof ConstantScalarType) {
+				$constScalars = $valueType->getConstantScalarValues();
+				if (count($constScalars) === 0) {
 					return null;
 				}
-				$arrayValues[] = $valueType->getValue();
+				$arrayValues[] = $constScalars;
 			}
 
-			$strings[] = new ConstantStringType(implode($separatorType->getValue(), $arrayValues));
+			$combinations = CombinationsHelper::combinations($arrayValues);
```

Comment:
Everywhere we use `CombinationsHelper::combinations` there's counting of the number of combinations beforehand, and ending early if it'd go over some limit, to prevent performance degradation.

Reply:
good catch, fixed